qiskit/transpiler/passes/routing/star_prerouting.py

```diff
@@ -0,0 +1,156 @@
+        if not star_sequences:
+            return dag
+        new_dag = dag.copy_empty_like()
+        sequence_starts = {x[0][0]: i for i, x in enumerate(star_sequences)}
+        processed_nodes = set()
+        qubit_mapping = {bit: index for index, bit in enumerate(dag.qubits)}
+        for node in dag.topological_op_nodes():
+            sequence_index = sequence_starts.get(node, None)
+            if sequence_index is not None:
+                index = sequence_index
+                sequence, center_node = star_sequences[index]
+                if len(sequence) == 2:
+                    for node in sequence:
+                        new_dag.apply_operation_back(
+                            node.op,
+                            _apply_mapping(node.qargs, qubit_mapping, dag.qubits),
+                            node.cargs,
+                        )
+                        processed_nodes.add(node)
+                    continue
+                swap_source = None
+                prev = None
+                for inner_node in sequence:
+                    if inner_node.qargs == prev:
+                        new_dag.apply_operation_back(
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
+                        )
+                        continue
+                    if swap_source is None:
+                        swap_source = center_node
+                        new_dag.apply_operation_back(
+                            inner_node.op,
+                            _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                            inner_node.cargs,
+                        )
+                        prev = inner_node.qargs
+                        processed_nodes.add(inner_node)
+                        continue
+                    new_dag.apply_operation_back(
+                        inner_node.op,
+                        _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                        inner_node.cargs,
+                    )
+                    processed_nodes.add(inner_node)
+                    new_dag.apply_operation_back(
+                        SwapGate(),
+                        _apply_mapping(inner_node.qargs, qubit_mapping, dag.qubits),
+                        inner_node.cargs,
+                    )
+                    # Swap mapping
+                    pos_0 = qubit_mapping[inner_node.qargs[0]]
+                    pos_1 = qubit_mapping[inner_node.qargs[1]]
+                    qubit_mapping[inner_node.qargs[0]] = pos_1
+                    qubit_mapping[inner_node.qargs[1]] = pos_0
+                    prev = inner_node.qargs
+            elif node in processed_nodes:
+                continue
+            else:
+                new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
+        return new_dag
```

Comment:
Well `ElidePermutations` hasn't merged yet, I'm still trying to figure out the layout issues there. But yeah we could in theory combine them, but yeah I like the idea of pairing them. I do wonder if there is a use case for this without adding the permutation, but doing this would work well. Maybe have it as a flag, like `reverse_permutation=True`, in the constructor to control whether we insert the `PermutationGate` or not?